### PR TITLE
[libc] Add Generic and NVPTX Sin Benchmark

### PR DESCRIPTION
**Pull request by jameshu15869**
This PR adds sin benchmarking for a range of values and on a pregenerated random distribution.

### CODE REVIEW

**Review by jhuber6 — COMMENTED** [review]
Neat, it's interesting to get very specific timings against the vendor implementations.

@lntue do you know if anything here is common with the math stuff you already have written?

File: libc/benchmarks/gpu/LibcGpuBenchmark.h [hunk 1]
```diff
@@ -102,6 +105,56 @@ class Benchmark {
...
+// |real value| <= 2^(max_exponent) * (1 + (random 52 bits) * 2^-52) <
+// 2^(max_exponent + 1)
+// The largest integer that can be stored in a double is 2^53
+static constexpr int MAX_EXPONENT = 52;
+static constexpr int RANDOM_INPUT_SIZE = 1024;
+static cpp::array<double, RANDOM_INPUT_SIZE> random_input;
+
+static double get_rand() {
+  using FPBits = LIBC_NAMESPACE::fputil::FPBits<double>;
+  uint64_t bits = LIBC_NAMESPACE::rand();
```
Comment (jhuber6) [inline libc/benchmarks/gpu/LibcGpuBenchmark.h:119]: If you want the results to be truly random we can probably just seed the RNG with the low bits of `get_processor_clock()`.

Comment (jhuber6) [reply libc/benchmarks/gpu/LibcGpuBenchmark.h:119]: I also wonder if we should make a reeantrant version of `rand` so we can have a state private to each thread instead of the shared state.

I.e.
`val = rand_r(&state)`

Comment (lntue) [reply libc/benchmarks/gpu/LibcGpuBenchmark.h:119]: I don't think for math function tests we really care too much about real random, as long as it shuffled enough

File: libc/benchmarks/gpu/LibcGpuBenchmark.cpp [hunk 2]
```diff
@@ -105,8 +106,34 @@ void print_results(Benchmark *b) {
...
+  return fp.get_val();
+}
+
+static void init_random_input() {
+  LIBC_NAMESPACE::srand(LIBC_NAMESPACE::gpu::processor_clock());
+  for (int i = 0; i < RANDOM_INPUT_SIZE; i++) {
+    random_input[i] = get_rand();
+  }
+}
+
```
Comment (jhuber6) [inline libc/benchmarks/gpu/LibcGpuBenchmark.cpp:131]: I figured this would be handled in the benchmark itself, like we do it before the latency calculation.

Comment (jameshu15869) [reply libc/benchmarks/gpu/LibcGpuBenchmark.cpp:131]: Ah, yeah that makes sense. I was thinking it might reduce duplication if we have to do the same thing for multiple benchmarks, but I think you're right about moving it to the benchmark itself

Comment (jameshu15869) [reply libc/benchmarks/gpu/LibcGpuBenchmark.cpp:131]: But now that I think about it, what was the way to share the array between all threads without using a global? I'm assuming that we want all threads to see the same random input, right?

Comment (jhuber6) [reply libc/benchmarks/gpu/LibcGpuBenchmark.cpp:131]: Can put it as a utility in the header.

Comment (jameshu15869) [reply libc/benchmarks/gpu/LibcGpuBenchmark.cpp:131]: Do you mean keep the array in the header but call `init_random_array()` from the benchmark itself?

Comment (jhuber6) [reply libc/benchmarks/gpu/LibcGpuBenchmark.cpp:131]: Yeah I just mean if you want a utility to get some random input, just make it a function in the header and have the benchmarks call it.

File: libc/benchmarks/gpu/LibcGpuBenchmark.cpp [hunk 3]
```diff
@@ -141,9 +141,9 @@ void print_header() {
 void Benchmark::run_benchmarks() {
   uint64_t id = gpu::get_thread_id();
 
-  if (id == 0)
+  if (id == 0) {
     print_header();
-
+  }
```
Comment (jhuber6) [inline libc/benchmarks/gpu/LibcGpuBenchmark.cpp:146]: No brackets.

File: libc/benchmarks/gpu/src/math/sin_benchmark.cpp [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+#endif
+
+constexpr double M_PI = 3.14159265358979323846;
+uint64_t get_bits(double x) {
+  return LIBC_NAMESPACE::cpp::bit_cast<uint64_t>(x);
+}
+
+constexpr int RANDOM_INPUT_SIZE = 256;
+
+// BENCHMARK() expects a function that with no parameters that returns a
```
Comment (jhuber6) [inline libc/benchmarks/gpu/src/math/sin_benchmark.cpp:22]: This should be in a utility header, but realisitically we just want to call `rand()` per-lane to give divergent inputs.

File: libc/benchmarks/gpu/LibcGpuBenchmark.h [hunk 5]
```diff
@@ -105,6 +109,56 @@ class Benchmark {
...
+  return fp.get_val();
+}
+
+template <size_t Size>
+static void init_random_double_input(cpp::array<double, Size> &values) {
+  LIBC_NAMESPACE::srand(LIBC_NAMESPACE::gpu::processor_clock());
+  for (int i = 0; i < Size; i++) {
+    values[i] = get_rand_double();
+  }
+}
```
Comment (jhuber6) [inline libc/benchmarks/gpu/LibcGpuBenchmark.h:135]: This isn't used anywhere, right? Can probably just drop it for now.

File: libc/benchmarks/gpu/LibcGpuBenchmark.h [hunk 6]
```diff
@@ -105,6 +109,56 @@ class Benchmark {
...
   }
 };
+
+// We want our random values to be approximately
+// |real value| <= 2^(max_exponent) * (1 + (random 52 bits) * 2^-52) <
+// 2^(max_exponent + 1)
+// The largest integer that can be stored in a double is 2^53
+static constexpr int MAX_EXPONENT = 52;
+
+static double get_rand_double() {
```
Comment (jhuber6) [inline libc/benchmarks/gpu/LibcGpuBenchmark.h:119]: We should probably template this because we'll want the same treatment for `sin` and `sinf` and `sinf16`.

File: libc/cmake/modules/LLVMLibCCompileOptionRules.cmake [hunk 7]
```diff
@@ -230,7 +230,7 @@ endfunction()
 function(_get_hermetic_test_compile_options output_var flags)
   _get_compile_options_from_flags(compile_flags ${flags})
   list(APPEND compile_options ${LIBC_COMPILE_OPTIONS_DEFAULT} ${compile_flags}
-       ${flags} -fpie -ffreestanding -fno-exceptions -fno-rtti)
+       ${flags} -fpie -ffreestanding -fno-exceptions -fno-rtti -save-temps)
```
Comment (jhuber6) [inline libc/cmake/modules/LLVMLibCCompileOptionRules.cmake:233]: Leftover debug (Also you might need to do `-Wl,--save-temps` if you're doing LTO since the backend is run inside the linker in that case and `-Wl` passes args to the linker).

File: libc/benchmarks/gpu/timing/amdgpu/timing.h [hunk 8]
```diff
@@ -98,7 +100,9 @@ template <typename F, typename T1, typename T2>
 
   auto result = f(arg1, arg2);
 
-  asm("v_or_b32 %[v_reg], 0, %[v_reg]\n" ::[v_reg] "v"(result) :);
+  asm("v_or_b32 %[v_reg], 0, %[v_reg]\n" ::[v_reg] "v"(
+          static_cast<uint32_t>(result))
+      :);
```
Comment (jhuber6) [inline libc/benchmarks/gpu/timing/amdgpu/timing.h:105]: Unrelated formatting change? I thin you last colon can be omitted here.

File: libc/benchmarks/gpu/LibcGpuBenchmark.h (hunk 5 above)
Comment (jameshu15869) [reply libc/benchmarks/gpu/LibcGpuBenchmark.h:135]: Oops, I missed this

File: libc/benchmarks/gpu/timing/amdgpu/timing.h (hunk 8 above)
Comment (jameshu15869) [reply libc/benchmarks/gpu/timing/amdgpu/timing.h:105]: Yeah, clang-format says that this line should be split into 2, but you're right about the last colon

File: libc/benchmarks/gpu/LibcGpuBenchmark.h (hunk 5 above)
Comment (jhuber6) [reply libc/benchmarks/gpu/LibcGpuBenchmark.h:135]: We might want to move this `srand()` call into the benchmark somewhere when we do other init stuff.

File: libc/benchmarks/gpu/LibcGpuBenchmark.h (hunk 6 above)
Comment (jameshu15869) [reply libc/benchmarks/gpu/LibcGpuBenchmark.h:119]: Since floats should have a different MAX_EXPONENT, how would you suggest we template this? Is there a nice way to switch the value of `MAX_EXPONENT`when using floats and doubles?

Comment (lntue) [reply libc/benchmarks/gpu/LibcGpuBenchmark.h:119]: `FPBits<T>::FRACTION_LEN`

Comment (jhuber6) [reply libc/benchmarks/gpu/LibcGpuBenchmark.h:119]: I forget the name of it, but it's standard `type_traits` stuff, see https://godbolt.org/z/4TGrvadYY.

File: libc/benchmarks/gpu/LibcGpuBenchmark.h [hunk 9]
```diff
@@ -113,27 +114,20 @@ class Benchmark {
...
-  using FPBits = LIBC_NAMESPACE::fputil::FPBits<double>;
-  uint64_t bits = LIBC_NAMESPACE::rand();
-  double scale = 0.5 + MAX_EXPONENT / 2048.0;
+template <typename T> static T get_rand_input() {
+  using FPBits = LIBC_NAMESPACE::fputil::FPBits<T>;
+
+  // Required to correctly instantiate FPBits for floats and doubles.
+  using RandType = typename cpp::conditional_t<(cpp::is_same_v<T, double>),
+                                               uint64_t, uint32_t>;
+  RandType bits = LIBC_NAMESPACE::rand();
```
Comment (jhuber6) [inline libc/benchmarks/gpu/LibcGpuBenchmark.h:123]: `rand()` always returns an int, so it's always 32-bit.

Comment (jameshu15869) [reply libc/benchmarks/gpu/LibcGpuBenchmark.h:123]: I think I was having some problems when trying to instantiate `FPBits` where it would only accept `uint64_t` for doubles and `uint32_t` for floats. Are you able to check if you can reproduce this locally?

Comment (jhuber6) [reply libc/benchmarks/gpu/LibcGpuBenchmark.h:123]: I wonder if that means we need to call `rand` twice to get an actual 64-bit value... @lntue might have some better insight here. Maybe we just need a better reentrant random function.

**Review by jhuber6 — COMMENTED** [review]
I guess we can put a call to `srand(gpu::processor_clock())` inside of the benchmark registration code.

**Comment by jameshu15869** [comment]
Ah, did you want to run `srand()` for every benchmark? Right now I have the call to `srand()` inside only the benchmark that uses it (e.g. `BM_RANDOM_INPUT` in `sin_benchmark.cpp`)

**Comment by jhuber6** [comment]
> Ah, did you want to run `srand()` for every benchmark? Right now I have the call to `srand()` inside only the benchmark that uses it (e.g. `BM_RANDOM_INPUT` in `sin_benchmark.cpp`). I was thinking that way each iteration would definitely get a different value, if that makes sense

`srand()` just initializes the seed, the implementation shares the global random state with every thread so we can just have a single one initialize it to something other than `1`.

**Comment by jameshu15869** [comment]
> `srand()` just initializes the seed, the implementation shares the global random state with every thread so we can just have a single one initialize it to something other than `1`.

Conceptually, should each individual benchmark (e.g. `Sin`, `NvSin`, etc.) call `srand()` once? Or should it just be called once for each invocation of our entire framework?

**Comment by jhuber6** [comment]
> > `srand()` just initializes the seed, the implementation shares the global random state with every thread so we can just have a single one initialize it to something other than `1`.
> 
> Conceptually, should each individual benchmark (e.g. `Sin`, `NvSin`, etc.) call `srand()` once? Or should it just be called once for each invocation of our entire framework?

Just call it once when we initialize the benchmarks, the sequence will then effectively be "random" whenever any other thread gets a value from the sequence.

**Comment by jameshu15869** [comment]
> Just call it once when we initialize the benchmarks, the sequence will then effectively be "random" whenever any other thread gets a value from the sequence.

I'm thinking about having thread 0 call `srand()` right before we start iterating through our benchmark list and running them (So it's only called once throughout the execution) - is that close to what you were thinking?

**Comment by jhuber6** [comment]
Pretty much like this.
```diff
diff --git a/libc/benchmarks/gpu/LibcGpuBenchmark.cpp b/libc/benchmarks/gpu/LibcGpuBenchmark.cpp
index a9a912538cd8..2574eb46f160 100644
--- a/libc/benchmarks/gpu/LibcGpuBenchmark.cpp
+++ b/libc/benchmarks/gpu/LibcGpuBenchmark.cpp
@@ -136,8 +136,10 @@ void print_header() {
 void Benchmark::run_benchmarks() {
   uint64_t id = gpu::get_thread_id();
 
-  if (id == 0)
+  if (id == 0) {
+    srand(gpu::processor_clock());
     print_header();
+  }
 
   gpu::sync_threads();
 ```

File: libc/benchmarks/gpu/timing/amdgpu/timing.h [hunk 10]
```diff
@@ -98,7 +99,9 @@ template <typename F, typename T1, typename T2>
...
```
Comment (jhuber6) [inline libc/benchmarks/gpu/timing/amdgpu/timing.h:104]: Is this colon required?

Comment (jameshu15869) [reply libc/benchmarks/gpu/timing/amdgpu/timing.h:104]: Oops, I swear I removed this already

**Review by jhuber6 — APPROVED** [review]
LG, thanks. Looking forward to getting this working for AMD and then testing a few more functions.
